### DISCUSSION TITLE
Sample parallel for each

### DISCUSSION THREAD

**Discussion by mremadevi** [elsa-workflows/elsa-core#5325] [Q&A]
Please provide a sample workflow using parallelforeach

**Comment by glime-ai** [comment]
Certainly! Below is a sample workflow that demonstrates the use of a `ParallelForEach` activity in Elsa Workflow. This example is a conceptual representation and assumes that you have a basic understanding of how to set up and execute workflows in Elsa.

```csharp
// Define a workflow with a ParallelForEach activity
public class ParallelForEachWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .StartWith<WriteLine>(x => x.Set(a => a.Text, "Starting parallel processing..."))
            .Then<ParallelForEach<string>>(parallelForEach =>
            {
                // Assuming we have a list of strings to process in parallel
                parallelForEach.Set(x => x.Items, new[] { "Task1", "Task2", "Task3" });
                parallelForEach.When(OutcomeNames.Iterate)
                    .Then<ProcessTask>() // Custom activity to process each item
                    .Then(parallelForEach); // Loop back to the ParallelForEach activity
            })
            .Then<WriteLine>(x => x.Set(a => a.Text, "Parallel processing completed."));
    }
}

// Custom activity to process each task
public class ProcessTask : Activity
{
    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        var taskName = context.GetInput<string>();
        Console.WriteLine($"Processing {taskName}...");
        // Simulate task processing by delaying for a short time
        Thread.Sleep(1000);
        Console.WriteLine($"{taskName} processed.");
        return Done();
    }
}
```

In this workflow, we start with a `WriteLine` activity to indicate the start of parallel processing. Then, we use the `ParallelForEach` activity to iterate over a list of strings, each representing a task. For each item in the list, we execute a custom activity `ProcessTask` that simulates processing of the task. After all tasks are processed in parallel, we finish with another `WriteLine` activity to indicate that the parallel processing is completed.

Please note that this is a simplified example for demonstration purposes. In a real-world scenario, you would replace the `ProcessTask` activity with your actual task processing logic.

To implement this workflow in Elsa, you would typically define these activities and the workflow in your application code and then use the Elsa services to execute the workflow.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
